adapters/madvertise/madvertise.go

```diff
@@ -0,0 +1,170 @@
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return &headers
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	placementID := ""
+	for _, imp := range request.Imp {
+		madvertiseExt, err := getImpressionExt(imp)
+		if err != nil {
+			return nil, []error{err}
+		}
+		if madvertiseExt.PlacementId != "" {
+			if placementID == "" {
+				placementID = madvertiseExt.PlacementId
+			} else if placementID != madvertiseExt.PlacementId {
+				return nil, []error{&errortypes.BadInput{
+					Message: "There must be only one placement ID",
+				}}
+			}
+		} else {
+			return nil, []error{&errortypes.BadInput{
+				Message: "The placement ID must not be empty",
+			}}
+		}
+		// request.Imp[idx].TagID = madvertiseExt.PlacementId
+	}
+
+	url, err := a.buildEndpointURL(placementID)
+	if err != nil {
+		return nil, []error{err}
+	}
+	var errors = make([]error, 0)
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     url,
+		Body:    requestJSON,
+		Headers: *getHeaders(request),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getImpressionExt(imp openrtb2.Imp) (*openrtb_ext.ExtImpMadvertise, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Bidder extension not provided or can't be unmarshalled",
+		}
+	}
+
+	var onetagExt openrtb_ext.ExtImpMadvertise
+	if err := json.Unmarshal(bidderExt.Bidder, &onetagExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+
+	return &onetagExt, nil
+}
+
+func (a *adapter) buildEndpointURL(placementID string) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{ZoneID: placementID}
+	return macros.ResolveMacros(a.endpointTemplate, endpointParams)
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bidMediaType, err := getMediaTypeForBid(request.Imp, bid)
```

Comment:
Please see our updated example in the docs for this loop. We now recommend the following to avoid reference bugs from the loop variable.

```
  for _, seatBid := range response.SeatBid {
    for _, bid := range seatBid.Bid {
      bid := bid // pin https://github.com/kyoh86/scopelint#whats-this
```

or another alternative is:

```
  for _, seatBid := range response.SeatBid {
    for i := range seatBid.Bid {
      bid := seatBid.Bid[i]
```

Reply:
done. on [master 674249d5] avoid reference bugs from the loop variable. bidResponse